### PR TITLE
Factorial and gamma functions

### LINKED ISSUE #50: Support factorials and the gamma function

### PR DESCRIPTION
**Pull request by Happypig375**
And fixes for nonfinite values for EDecimal extensions

Resolves #50

### CODE REVIEW

**Comment by WhiteBlackGoose** [comment]
Oof that's a huge PR. It seems okay overall, but I need to dig into it a bit deeper. Sorry for the late response

**Comment by WhiteBlackGoose** [comment]
Don't we need some kind of [inverse Gamma function](https://mathoverflow.net/questions/12828/inverse-gamma-function) for equation solving?

**Comment by Happypig375** [comment]
To implement it, we need the Lambert W function (as noted in https://github.com/Happypig375/AngouriMath/blob/d4a7da91c230ed925775f2e11b2093012a0c9c8b/AngouriMath/Functions/Algebra/EquationSolver/Analytical/AnalyticalSolver.cs). I couldn't find an implementation for arbitrary precision evaluation of this function.

**Comment by Happypig375** [comment]
Seems like we need to call into the Newton Solver for evaluating this function: https://www.quora.com/How-is-the-Lambert-W-Function-computed?share=1

**Comment by WhiteBlackGoose** [comment]
So we need some kind of inverse for Gamma and Gamma's derivative.

File: AngouriMath/Functions/Algebra/Derivation.cs [hunk 1]
```diff
@@ -220,4 +223,15 @@ public static Entity Derive(List<Entity> args, VariableEntity variable)
...
         }
     }
+    internal static partial class Factorialf
+    {
+        // (x!)' = Γ(x + 1) polygamma(0, x + 1)
+        public static Entity Derive(List<Entity> args, VariableEntity variable)
+        {
+            MathFunctions.AssertArgs(args.Count, 1);
+            var a = args[0];
+            // TODO: Implementation of symbolic gamma function and polygamma functions needed
```
[the remark this answers is not in the document]
Comment (Happypig375) [reply AngouriMath/Functions/Algebra/Derivation.cs:233]: That derivative is limited to positive integers. We need a derivative that works for all complex numbers - hence the polygamma functions.

**Pull request merged**